### PR TITLE
Set up a basic automated e2e smoke test

### PR DESCRIPTION
**Pull request by qobilidop**
## Summary

This PR sets up a basic automated e2e smoke test:

- Automated means the whole testing is runnable with a single command.
- e2e means it exercises the whole P4 DPDK stack including the p4c compiler (with DPDK backend), DPDK software switch, TDI runtime API, and some other libraries provided in this repo.

## Background

In our current experimental usage of the P4 DPDK stack, we sometimes identify issues that could be caught early by a simple e2e smoke test. We think adding these tests to this public repo, and setting up GitHub CI to run them automatically, could be a useful addition to the existing unit tests.

## Demo

Build a Docker image with the P4 DPDK stack (can take up to hours):

```console
$ ./tools/docker_build.sh
```

Run the smoke test in a container:

```console
$ ./tools/docker_run_smoke_test.sh
INFO:root:Run command:
    p4c-dpdk --arch pna --p4runtime-files e2e-test/suites/pna/basicfwd/p4c_gen/p4info.txt --bf-rt-schema e2e-test/suites/pna/basicfwd/p4c_gen/bfrt.json --tdi e2e-test/suites/pna/basicfwd/p4c_gen/tdi.json --context e2e-test/suites/pna/basicfwd/p4c_gen/context.json -o e2e-test/suites/pna/basicfwd/p4c_gen/config.spec e2e-test/suites/pna/basicfwd/main.p4

INFO:root:Using SDE: /home/sde
INFO:root:Using SDE_INSTALL: /home/sde/install
Node Pages Size Total
0    16    2Mb    32Mb
1    16    2Mb    32Mb

Hugepages mounted on /dev/hugepages
INFO:root:Run command:
    sudo -E PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin LD_LIBRARY_PATH=/home/sde/install/lib:/home/sde/install/lib64:/home/sde/install/lib/x86_64-linux-gnu /home/sde/install/bin/bf_switchd --install-dir /home/sde/install --conf-file /home/work/e2e-test/suites/pna/basicfwd/conf_bf_switchd.json

INFO:root:PID: 34
INFO:root:Run command:
    /home/sde/install/bin/bfshell -f /home/work/e2e-test/suites/pna/basicfwd/cmds_bfshell.py

INFO:root:PID: 50
net.ipv6.conf.all.disable_ipv6 = 1
net.ipv6.conf.default.disable_ipv6 = 1
net.ipv6.conf.lo.disable_ipv6 = 1
.
Sent 1 packets.
..
Sent 1 packets.
.
----------------------------------------------------------------------
Ran 2 tests in 2.321s

OK
```

For more details, see `e2e-test/README.md`.

## Current status and further plans

The test added in this PR is very basic, but shall serve as a good starting point for test organization.

@lzhzero and I will keep working on adding more tests in follow-up PRs, and reporting issues if we encountered any, until we get a fully GitHub CI automated representative sets of e2e tests.

### CODE REVIEW

**Comment by qobilidop** [comment]
One specific question I have now is how to best set up the Docker images? I could think of several different options, and have implemented one of them. But there could be reasons to prefer other options.

1. (This one is implemented) A Docker image based on the existing [p4lang/p4c image](https://hub.docker.com/r/p4lang/p4c) built using this [Dockerfile](https://github.com/p4lang/p4c/blob/main/Dockerfile) in the p4c repo continuously.
2. We could build p4c from source within the Docker image. One benefit is the possibility to pin a specific version of p4c to test against instead of always using the latest one.
3. We could have separate p4c and p4-dpdk Docker images, because these 2 stacks don't depend on each other. Then in testing, we can use the p4c container to compile the P4 code, and the p4-dpdk container to run the test against compiler outputs.

Any inputs on this are appreciated!

**Comment by qobilidop** [comment]
I'm encountering a new issue that could use some help.

## Short description

The `add_with_xxx` command fails with:

```
TdiTableError: Error: table_entry_add failed on table pipe.MainControl.forwarding. [Object not found]
```

Relevant logs shown in `bf_switchd`:

```
MatchSpec:
priority = 0
num_valid_match_bits:48
num_match_bytes:6
match_value_bits: 0x0000c0a80166
match_mask_bits: 0xffffffffffff
ActionSpec:
action_bmap:01
num_valid_action_data_bits:64
num_action_data_bytes:8
action_data: 0x0000000200000002
get_table_id failed at 693
get_table_id failed at 714 end
2023-06-02 23:54:05.700968 BF_PIPE ERROR - dpdk table id get for table forwarding failed
2023-06-02 23:54:05.701086 BF_PIPE ERROR - not able get table metadata for table forwarding
2023-06-02 23:54:05.701134 BF_PIPE ERROR - dal_table_ent_add failed
```

## How to reproduce?

Use the latest commit in this PR. Follow the instructions in `e2e-test/README.md`.

For the convenience of a quick look, below I copied the `bfshell` commands file I used:

```python
tdi_python

# Add ports

for port_id in range(4):
    tdi.port.port.add(
        DEV_PORT=port_id,
        PORT_TYPE="BF_DPDK_TAP",
        PORT_DIR="PM_PORT_DIR_DEFAULT",
        PORT_IN_ID=port_id,
        PORT_OUT_ID =port_id,
        PIPE_IN="pipe",
        PIPE_OUT="pipe",
        MEMPOOL="MEMPOOL0",
        PORT_NAME=f"TAP{port_id}",
        MTU=1500
    )

# Add entries

from netaddr import IPAddress

control = tdi.main.pipe.MainControl
table = control.forwarding
table.add_with_mark_and_forward(
    dst_addr=IPAddress('192.168.1.101'),
    marker=0x01,
    port=1
)
table.add_with_mark_and_forward(
    dst_addr=IPAddress('192.168.1.102'),
    marker=0x02,
    port=2
)
```

The error occurs on the last two `table.add_with_mark_and_forward` calls.

**Comment by jafingerhut** [comment]
Are you able to create at least one simple test that passes, that is not blocked by the issue you have found and reported in this comment?  https://github.com/p4lang/p4-dpdk-target/pull/102#issuecomment-1574461454

If so, it might be nice to focus on creating that first simple passing test first, and file a separate issue for the bug you have found.

**Comment by qobilidop** [comment]
Thanks Andy for your suggestion! I'll try to bypass this and come to a working test first.

**Comment by jafingerhut** [comment]
@satish153 @swaroopsarma @jamescchoi Can you suggest someone who can (a) review this?  and (b) do whatever approval is required for the Github "1 workflow awaiting approval" message I see on this conversation?

**Comment by qobilidop** [comment]
Thanks @fruffy for offering help! That would be great!

I just tried building the Docker image again. It takes 45 min on my computer, and the final image is 4.78 GB. I feel the building time might be too long, and the image size too large to be usable in CI.

Without a Docker container, as long as all the dependencies are installed, the testing scripts shall work. Maybe that's an easier solution on the CI side.

I still think providing a Dockerfile is useful as a reproducible environment to try things out locally, even if it's not used in the CI. I feel the best middle ground might be what's been done in the p4c repo, that the Dockerfile is simply invoking the ci-build.sh script, which is also used in CI to build required dependencies.

One challenge I faced in setting up dependencies is how to best bring in p4c. Building p4-dpdk-target itself doesn't require p4c, but in the e2e tests p4c is needed. My current solution is to use the public p4lang/p4c image as the base image, so that p4c is automatically available. This is probably not ideal, because the p4c version is not fixed, but sometimes that might matter. If we do want to fix p4c version in testing, then it seems building p4c from source is the only option?

I'd like to continue with more detailed discussions on this and some other issues, but I feel getting them all resolved in this PR might be too much. So, I'm proposing limiting the scope of this PR to a single working test that is automated locally, and expect more work to be done before it is fully automated in CI. Hopefully this could also make reviewing this PR easier.

**Comment by swaroopsarma** [comment]
Just rebased the pull request

File: src/pipe_mgr/shared/dal/dpdk/dal_init.c
Comment (qobilidop) [reply src/pipe_mgr/shared/dal/dpdk/dal_init.c:129]: Sure, will do.

Comment (qobilidop) [reply src/pipe_mgr/shared/dal/dpdk/dal_init.c:129]: I have done more experiments, and this issue is becoming very mysterious, almost "quantom mechanical". To describe my observations, let me first introduce 2 variations of the code here:

V1:
```cpp
		status = system(buffer);
		LOG_ERROR("%s line:%d  Returned status: %d\n",
			  __func__, __LINE__, status);
		if (status) {
			LOG_ERROR("%s line:%d  Cannot generate %s file\n",
				  __func__, __LINE__, so_filepath);
			return BF_INTERNAL_ERROR;
		}
```

V2:
```cpp
		status = system(buffer);
		if (status) {
			LOG_ERROR("%s line:%d  Returned status: %d\n",
				  __func__, __LINE__, status);
			LOG_ERROR("%s line:%d  Cannot generate %s file\n",
				  __func__, __LINE__, so_filepath);
			return BF_INTERNAL_ERROR;
		}
```

- I'm using `LOG_ERROR` here instead of `LOG_TRACE`, because otherwise the log message won't show up in `p4_driver.log`. I suspect this is a bug.

Observations:
- With V1, the logged message is `Returned status: 0`, and the test will pass.
- With V2, the logged message is `Returned status: -1`, and the test will fail.

I cannot make any sense of these observations. I checked the code that `LOG_ERROR` won't modify the value of `status`. Then how could the value of `status` change depending on whether it is logged or not?

**I wonder if anyone other than me can reproduce these observations?**

If you have built the project locally, just rebuild it with V1 or V2, and run the following command:
```console
# Have to make sure these files generated previously are removed,
# or the error won't show up in V2.
rm /tmp/p4.*
python3 e2e-test/tools/test.py e2e-test/suites/pna/basicfwd
```

After it finished, check the file `e2e-test/suites/pna/basicfwd/log/bf_switchd/p4_driver.log` to see the log messages.

Comment (swaroopsarma) [reply src/pipe_mgr/shared/dal/dpdk/dal_init.c:129]: is the so generated even when the return value is -1 ? The only difference i can think of in between v1 & v2 is that the 
 implicit sync operation wehenever the print executes.

Comment (qobilidop) [reply src/pipe_mgr/shared/dal/dpdk/dal_init.c:129]: Yes, the so file is generated even when the return value is -1.

**Comment by jafingerhut** [comment]
@swaroopsarma Can you take another look at this to see if it looks good to you now?  If I understand correctly, with the latest commit (number 11) on this PR, there are only new log messages output in the dal_init.c code, but no other functional changes.

File: src/pipe_mgr/shared/dal/dpdk/dal_init.c
Comment (qobilidop) [reply src/pipe_mgr/shared/dal/dpdk/dal_init.c:129]: I decided to just go with V1 for now and leave the more subtle issue of logging influencing return value for future discussion maybe in its own issue.

Comment (qobilidop) [reply src/pipe_mgr/shared/dal/dpdk/dal_init.c:129]: I did some more experiments today, and this V1/V2 difference is not reproducible. I did observe the same behaviors in some trials, but eventually V1 also gives -1 return status.

I've adopted another workaround to solve this issue, which is to check .so file existence instead of status here. I know this is not an ideal solution, but maybe we can move on with this temporary solution and try to solve it later? What do you think?

**Comment by qobilidop** [comment]
@jafingerhut Thanks for the ping and sorry about the premature pushes. I added 2 more commits and now they should be ready for another review.

Also thank you @swaroopsarma for your previous review. Wish to hear more from you soon.

File: src/pipe_mgr/shared/dal/dpdk/dal_init.c
Comment (qobilidop) [reply src/pipe_mgr/shared/dal/dpdk/dal_init.c:129]: Some further thoughts on this. Since the .so file is later used in `rte_swx_pipeline_build_from_lib(...)`, the validity of the file is checked there.

**Comment by chrispsommers** [comment]
> This is probably not ideal, because the p4c version is not fixed, but sometimes that might matter. If we do want to fix p4c version in testing, then it seems building p4c from source is the only option?

HI @qobilidop This is easy to solve, you can docker pull a specific SHA of any docker image instead of using e.g. `:latest` or `:stable`, e.g. here is the version of p4c:latest on 2022-08-19. You can navigate dockerhub and look at the tag versions, [see example]( https://hub.docker.com/layers/p4lang/p4c/latest/images/sha256-f334bfa3f3f5fc92c3935dc88dc2dbd35a17b1d332fcb98041d5a660c666b2dc?context=explore) then get the SHA from there. You can pull or run this image directly using `@sha<sha code>:` instead of the normal `tag`, for example:
```
docker pull p4lang/p4c@sha256:f334bfa3f3f5fc92c3935dc88dc2dbd35a17b1d332fcb98041d5a660c666b2dc
```
This is how you can pin a version. Just document in any scripts what the SHA corresponds to, e.g. `# This is same as p4c:latest on 2023-06-23`

BTW you can also use this to build your own dockerfiles using a specific base image, I do this in the SONiC-DASH project. Here's an example Dockerfile which I base on a specific p4c docker imaghe. I then make a new docker image which is smaller by throwing away unneeded backends. Then I publish this smaller custom p4c image to another container registry which a CI pipeline or user dev environment can pull down. I saved a few hundred MB this way. https://github.com/sonic-net/DASH/blob/main/dash-pipeline/dockerfiles/Dockerfile.p4c-bmv2

**Comment by chrispsommers** [comment]
To avoid long docker build times, why don't you automate the docker build and publish workflow similar to many other p4lang repos, e.g. https://github.com/p4lang/p4c/blob/main/.github/workflows/ci-container-image.yml ?

Everything gets built and published to Dockerhub daily. This will save hours, and perhaps attract more users because the initial user experience can be a deal-breaker. Strive for 5 Minutes TFD (Tine for First Dopamine).

**Comment by qobilidop** [comment]
Thanks @chrispsommers! Those are great suggestions. I didn't realize it could be that easy. The example Dockerfile you linked is especially useful. I see there is also a `Dockerfile.p4c-dpdk`. I'll learn from those examples and improve the Dockerfile in this PR.

As to publishing Docker images using GitHub CI automatically, I completely agree with you that this would be a great thing to have. I'm just not sure if I am the right person to set it up. Because it seems this would require more permissions (both on GitHub and DockerHub) than I currently have. I'm willing to do the work, but it might be easier for someone already having those permissions to set this up in a follow-up PR. @swaroopsarma @jafingerhut @fruffy What do you think?

**Comment by chrispsommers** [comment]
>  I'm just not sure if I am the right person to set it up. Because it seems this would require more permissions (both on GitHub and DockerHub) than I currently have. I'm willing to do the work, but it might be easier for someone already having those permissions to set this up in a follow-up PR. @swaroopsarma @jafingerhut @fruffy What do you think?

@qobilidop while developing it in your forked dev branch, you can take a few temporary measures to develop it as far as possible w/o involving others, e.g.:
- comment out the final login/publish steps in the .yml CI file, this way you can debug the build process itself
- change the Docker repo to a personal one on Dockerhub, I did this for DASH and we actually used it for several months in production until we migrated the docker registry to ACR.

Actually puling the trigger to publish to p4lang/p4-dpdk docker registry would probably require the p4-dpdk account admin to install the dockerhub account credentials in the repo secrets (unless they are global for all or p4lang GitHub site).

**Comment by qobilidop** [comment]
Got it. I'll try these initial steps then. Thanks again @chrispsommers!

**Comment by chrispsommers** [comment]
One more piece of advice - any changes you make to this branch will trigger CI actions in the p4lang/p4-dpdk GitHub pipeline, because the PR is already opened. If you want to try out a new CI script or other potentially messy experiments without sharing with the entire world, you can make a scratch branch off of it in your workspace (`git checkout -b some-name`) and push to your own repo. There you can do messy experiments while you're trying CI actions, etc. in relative privacy. You must enable Git Actions in your repo, they're off by default. Then you can merge back to the PR branch and the same actions will get triggered in the PR branch. I wish I'd thought of this before a few long, embarrassing sessions. :)

**Comment by fruffy** [comment]
> Thanks @chrispsommers! Those are great suggestions. I didn't realize it could be that easy. The example Dockerfile you linked is especially useful. I see there is also a `Dockerfile.p4c-dpdk`. I'll learn from those examples and improve the Dockerfile in this PR.
> 
> As to publishing Docker images using GitHub CI automatically, I completely agree with you that this would be a great thing to have. I'm just not sure if I am the right person to set it up. Because it seems this would require more permissions (both on GitHub and DockerHub) than I currently have. I'm willing to do the work, but it might be easier for someone already having those permissions to set this up in a follow-up PR. @swaroopsarma @jafingerhut @fruffy What do you think?

Iiirc there is no formal process. The docker images are maintained on a voluntary basis. I believe @antoninbas has access to the p4lang Dockerhub.

**Comment by qobilidop** [comment]
@chrispsommers I've adopted the multi-stage builds technique from the examples you linked to. It helps reducing docker image size from 4.78GB to 1GB, and building time from 45min to 13min. Very impressive!

For the docker image CI setup, I still plan to do the work, but after second thoughts, I again propose to leave it to a follow-up PR, so we can leave this already-long PR more focused (and potentially easier to review). I can work on this in my fork without waiting for this PR to be merged. So that's not a blocker.

@swaroopsarma I've marked this PR as ready for review again. The main change needing your attention (based on your last review comments) is the diff in `dal_init.c`. I have come up with a new workaround, and would like to know if it is good enough this time.

@jafingerhut @fruffy The bmv2 ptf testing work you mentioned in another meeting sounds pretty relevant here. I'm interested to learn how it works if you can share a link, and see if I can help with integrating with it in a follow-up PR.

Thanks everyone!

**Comment by fruffy** [comment]
> @jafingerhut @fruffy The bmv2 ptf testing work you mentioned in another meeting sounds pretty relevant here. I'm interested to learn how it works if you can share a link, and see if I can help with integrating with it in a follow-up PR.

I believe Andy did a writeup [here](https://github.com/jafingerhut/p4-guide/blob/master/ipdk/23.01/README-install-ipdk-networking-container-ubuntu-20.04-and-test.md). The main advantage is the direct use of Protobuf messages. I do not think this setup requires the shell. Correct me if I am wrong, Andy. 

I am also working with a student at NYU (@Hoooao) on an automated setup for p4c, which uses P4Testgen to generate PTF tests for the SoftNIC and runs them. We were thinking of using your automated workflow and then combining it with the PTF tests setup. But this is still work in progress.

**Comment by chrispsommers** [comment]
> I am also working with a student at NYU (@Hoooao) on an automated setup for p4c, which uses P4Testgen to generate PTF tests for the SoftNIC and runs them. We were thinking of using your automated workflow and then combining it with the PTF tests setup. But this is still work in progress.

@fruffy Interesting coincidence, I'd just messaged @jafingerhut about a proposal to include p4tools in all nightly p4c docker build/publish CI runs. It would make using p4testgen more convenient, just pull a pre-built docker. I'll file a PR shortly.

**Comment by qobilidop** [comment]
Thanks @fruffy for the information. That's great to know! This kind of test infra surely will be very useful. I look forward to you and @Hoooao's work. I'll read Andy's writeup for my own learning then.

**Comment by chrispsommers** [comment]
@fruffy I've submitted PR https://github.com/p4lang/p4c/pull/4049 to incorporate p4tools (including p4testgen) into Dockerfile builds which will eliminate the need to build from scratch unless you're developing the backend itself. Consumers can just `docker pull p4lang/p4c` and use it. It needs a review.

**Comment by jafingerhut** [comment]
> I believe Andy did a writeup [here](https://github.com/jafingerhut/p4-guide/blob/master/ipdk/23.01/README-install-ipdk-networking-container-ubuntu-20.04-and-test.md). The main advantage is the direct use of Protobuf messages. I do not think this setup requires the shell. Correct me if I am wrong, Andy.

My PTF tests use a Python library called p4runtime-shell just out of personal choice: https://github.com/p4lang/p4runtime-shell

Any method of creating+sending, or receiving+parsing, the required P4Runtime Protobuf messages in Python will work.

File: src/pipe_mgr/shared/dal/dpdk/dal_init.c
Comment (lzhzero) [inline src/pipe_mgr/shared/dal/dpdk/dal_init.c:114]: I think the -1 return value could be caused by race condition of snprintf.
running the original code with gdb will get return value of 0.

Comment (qobilidop) [reply src/pipe_mgr/shared/dal/dpdk/dal_init.c:114]: Ahh, that can explain why this occurs probabilistically! Do you know how to avoid the race condition here?

**Review by swaroopsarma — APPROVED** [review]
looks good

File: src/pipe_mgr/shared/dal/dpdk/dal_init.c
Comment (swaroopsarma) [reply src/pipe_mgr/shared/dal/dpdk/dal_init.c:114]: track this issue separately

**Comment by swaroopsarma** [comment]
@qobilidop  and @jafingerhut  can we merge this PR ?

**Comment by qobilidop** [comment]
I'm okay with merging it. And thanks for reviewing!

**Comment by jafingerhut** [comment]
@swaroopsarma I have no objections to merging it.  Please do not count on me for a technical review of these changes -- that would be a job best requested of others who know this repo contents better than I do.

I am not sure if you were asking me to do perform the merge, but if you were, I do not have permissions on this repository to do that.

**Pull request merged**
